### DISCUSSION TITLE
Can't Generate a Random BigInteger in a Specified Range

### DISCUSSION THREAD

**Discussion by YBCA7** [dotnet/runtime#98884] [Q&A] [answered]
It seems that the code below has no problem. However, it usually returns a number out of range.
```csharp
using System;
using System.Numerics;
using System.Security.Cryptography;

private BigInteger EncryptedRandBigInt(BigInteger min, BigInteger max, RNGCryptoServiceProvider r)
{
    BigInteger range = max - min;
    byte[] buffer = new byte[(int)range.BitLength /8 + 1];
    do { r.GetBytes(buffer); } while (new BigInteger(buffer) > range);
    return new BigInteger(buffer) + min;
}
```

**Comment by Clockwork-Muse** [comment, marked as answer]
Because %50 of the time the generated bytes will result in a negative number, due to the high bit of the last byte being set.  You have to explicitly set the high bit to zero to force it to be positive.

See also [this explanatory SO answer](https://stackoverflow.com/a/48855115/812837) (Although use of some more modern APIs may allow you to cut down on the allocations).
